sklearn/mixture/base.py

```diff
@@ -0,0 +1,425 @@
+"""Base class for mixture models."""
+
+# Author: Wei Xue <xuewei4d@gmail.com>
+# Modified by Thierry Guillemot <thierry.guillemot.work@gmail.com>
+
+from __future__ import print_function
+
+import warnings
+from abc import ABCMeta, abstractmethod
+from time import time
+
+import numpy as np
+
+from .. import cluster
+from ..base import BaseEstimator
+from ..base import DensityMixin
+from ..externals import six
+from ..exceptions import ConvergenceWarning
+from ..utils import check_array, check_random_state
+from ..utils.extmath import logsumexp
+
+
+def _check_shape(param, param_shape, name):
+    """Validate the shape of the input parameter 'param'.
+
+    Parameters
+    ----------
+    param : array
+
+    param_shape : tuple
+
+    name : string
+    """
+    param = np.array(param)
+    if param.shape != param_shape:
+        raise ValueError("The parameter '%s' should have the shape of %s, "
+                         "but got %s" % (name, param_shape, param.shape))
+
+
+def _check_X(X, n_components=None, n_features=None):
+    """Check the input data X.
+
+    Parameters
+    ----------
+    X : array-like, (n_samples, n_features)
+
+    n_components : int
+
+    Returns
+    -------
+    X : array, (n_samples, n_features)
+    """
+    X = check_array(X, dtype=[np.float64, np.float32])
+    if n_components is not None and X.shape[0] < n_components:
+        raise ValueError('Expected n_samples >= n_components'
+                         'but got n_components = %d, n_samples = %d'
+                         % (n_components, X.shape[0]))
+    if n_features is not None and X.shape[1] != n_features:
+        raise ValueError("Expected the input data X have %d features, "
+                         "but got %d features"
+                         % (n_features, X.shape[1]))
+    return X
+
+
+class MixtureBase(six.with_metaclass(ABCMeta, DensityMixin, BaseEstimator)):
+    """Base class for mixture models.
+
+    This abstract class specifies an interface for all mixture classes and
+    provides basic common methods for mixture models.
+    """
+
+    def __init__(self, n_components, tol, reg_covar,
+                 max_iter, n_init, init_params, random_state, warm_start,
+                 verbose, verbose_interval):
+        self.n_components = n_components
+        self.tol = tol
+        self.reg_covar = reg_covar
+        self.random_state = random_state
+        self.max_iter = max_iter
+        self.n_init = n_init
+        self.init_params = init_params
+        self.warm_start = warm_start
+        self.verbose = verbose
+        self.verbose_interval = verbose_interval
+
+        self.n_features = None
+        self.converged_ = False
+        self.n_iter_ = 0
+        self.initialized_ = False
```

Comment:
init method should only do the assignments. This second block should be done in fit